lib/db_connection.ex

```diff
@@ -738,45 +736,134 @@ defmodule DBConnection do
     case checkout(pool, nil, opts) do
-      {:ok, conn, _, _} ->
-        try do
-          fun.(conn)
-        after
-          checkin(conn, opts)
-        end
-      {:error, err, _} ->
+      {:ok, conn, _conn_state, _meter} ->
+        {:ok, conn}
+      {:error, err, _meter} ->
+        {:error, err}
+      {kind, reason, stack, _meter} ->
+        :erlang.raise(kind, reason, stack)
+    end
+  end
+
+  @doc """
+  Acquire a lock on a connection.
+
+  Returns the connection reference on success, and raises an exception on error.
+  See `checkout/2`.
+  """
+  @spec checkout!(pool :: GenServer.t, opts :: Keyword.t) :: t
+  def checkout!(pool, opts \\ []) do
+    case checkout(pool, nil, opts) do
+      {:ok, conn, _conn_state, _meter} ->
+        conn
+      {:error, err, _meter} ->
         raise err
-      {kind, reason, stack, _} ->
+      {kind, reason, stack, _meter} ->
         :erlang.raise(kind, reason, stack)
     end
   end
 
+
   @doc """
-  Acquire a lock on a connection and run a series of requests inside a
-  transaction. The result of the transaction fun is return inside an `:ok`
-  tuple: `{:ok, result}`.
+  Release a lock on a connection.
 
-  To use the locked connection call the request with the connection
-  reference passed as the single argument to the `fun`. If the
-  connection disconnects all future calls using that connection
-  reference will fail.
+  Returns `:ok` on success, otherwise `{:error, exception}` on error.
+
+  The pool may support options.
+
+  ### Example
+
+      {:ok, conn} = DBConnection.checkout(pool)
+      try do
+        DBConnection.execute(conn, "SELECT * FROM table", [])
+      after
+        DBConnection.checkin(conn)
+      end
+  """
+  @spec checkin(t, opts :: Keyword.t) :: :ok | {:error, Exception.t}
+  def checkin(%DBConnection{} = conn, opts \\ []) do
+    case delete_conn(conn) do
+      {:ok, conn_state} ->
+        checkin(conn, conn_state, opts)
+      {:error, _} = error ->
+        error
+    end
+  end
+
+  @doc """
+  Release a lock on a connection.
 
-  `run/3` and `transaction/3` can be nested multiple times. If a transaction is
-  rolled back or a nested transaction `fun` raises the transaction is marked as
-  failed. All calls except `run/3`, `transaction/3`, `rollback/2`, `close/3` and
-  `close!/3` will raise an exception inside a failed transaction until the outer
-  transaction call returns. All `transaction/3` calls will return
-  `{:error, :rollback}` if the transaction failed or connection closed and
-  `rollback/2` is not called for that `transaction/3`.
+  Returns `:ok` on sucess, otherwsie raises an exception on error.
+
+  See `checkin/2`.
+  """
+  @spec checkin!(t, opts :: Keyword.t) :: :ok
+  def checkin!(conn, opts \\ []) do
+    case checkin(conn, opts) do
+      :ok ->
+        :ok
+      {:error, err} ->
+        raise err
+    end
+  end
+
+  @doc """
+  Begin a transaction.
+
+  Return `{:ok, result}` on sucess or `{:error, exception}` if there was an
```

Comment:
This approach doesn't work nicely with logging, because these requests will be logged when no query is occurring, and if either of those are returned it will crash loggers.

There are also issues with some DBs. PostgreSQL ignores (but warns on) multiple start transactions, whereas MySQL will do an implicit commit and start a fresh transaction. Also PostgreSQL will implicitly turn committing transaction into rolling back transaction in an aborted state. This means providing the safe/ecto style is also very opinionated, and doesn't fit any database behaviour.